rust/src/witness_builder.rs

```diff
@@ -0,0 +1,393 @@
+        let redeemers = self.redeemers.iter().map(|key| format!("Redeemer:{}-{}", hex::encode(key.tag().to_bytes()), key.index().to_str())).collect::<Vec<String>>().join(",");
+
+        [vkeys, bootstraps, native_scripts, plutus_scripts, plutus_data, redeemers].iter().filter(|msg| msg.len() > 0).cloned().collect::<Vec<String>>().join("\n")
+    }
+
+    pub (crate) fn len(&self) -> usize {
+        self.vkeys.len() +
+            self.bootstraps.len() +
+            self.native_scripts.len() +
+            self.plutus_scripts.len() +
+            self.plutus_data.len() +
+            self.redeemers.len()
+    }
+
+    pub fn new() -> Self {
+        Self {
+            vkeys: HashSet::new(),
+            bootstraps: HashSet::new(),
+            native_scripts: HashSet::new(),
+            plutus_scripts: HashSet::new(),
+            plutus_data: HashSet::new(),
+            redeemers: HashSet::new(),
+        }
+    }
+}
+
+/// Builder de-duplicates witnesses as they are added 
+#[wasm_bindgen]
+#[derive(Clone)]
+pub struct TransactionWitnessSetBuilder {
+    // See Alonzo spec section 3.1 which defines the keys for these types
+    vkeys: HashMap<Vkey, Vkeywitness>,
+    bootstraps: HashMap<Vkey, BootstrapWitness>,
+    native_scripts: HashMap<ScriptHash, NativeScript>,
+    plutus_scripts: HashMap<ScriptHash, PlutusScript>,
+    plutus_data: HashMap<DataHash, PlutusData>,
+    redeemers: HashMap<RedeemerWitnessKey, Redeemer>,
+
+    // signatures that need to be added for the build function to succeed
+    required_sigs: RequiredSignatureSet,
+}
+
+#[wasm_bindgen]
+impl TransactionWitnessSetBuilder {
+    pub fn add_vkey(&mut self, vkey: &Vkeywitness) {
+        self.vkeys.insert(vkey.vkey(), vkey.clone());
+    }
+
+    pub fn add_bootstrap(&mut self, bootstrap: &BootstrapWitness) {
+        self.bootstraps.insert(bootstrap.vkey(), bootstrap.clone());
+    }
+
+    pub fn add_native_script(&mut self, native_script: &NativeScript) {
+        self.native_scripts.insert(native_script.hash(ScriptHashNamespace::NativeScript), native_script.clone());
+    }
+
+    pub fn add_plutus_script(&mut self, plutus_script: &PlutusScript) {
+        // TODO: don't assume PlutusV1 and instead somehow calculate this
+        self.plutus_scripts.insert(plutus_script.hash(ScriptHashNamespace::PlutusV1), plutus_script.clone());
+    }
+
+    pub fn add_plutus_datum(&mut self, plutus_datum: &PlutusData) {
+        self.plutus_data.insert(hash_plutus_data(&plutus_datum), plutus_datum.clone());
+    }
+
+    pub fn add_redeemer(&mut self, redeemer: &Redeemer) {
+        self.redeemers.insert(
+            RedeemerWitnessKey::new(&redeemer.tag(), &redeemer.index()),
+            redeemer.clone()
+        );
+    }
+
+    pub fn set_required_sigs(&mut self, required_sigs: &RequiredSignatureSet) {
+        self.required_sigs = required_sigs.clone()
+    }
+
+    pub fn get_required_sigs(&self) -> RequiredSignatureSet {
```

Comment:
I'm not sure if we want to expose this getter since it's misleading (it returns a copy of a mutable object -- not a reference)

Reply:
I ended up removing it since I feel it's just a foot-gun